src/lib/tests/tests.cpp

```diff
@@ -1,6 +1,399 @@
+  fx4.storeU(result);
+  for (size_t i = 0; i < 4; i++) {
+    EXPECT_FLOAT_EQ(result[i], 1.0F);
+  }
+  delete[] result;
+}
+
+/* __m128 requires 16 bit alignment */
+TEST(SimdFloatX4, AlignedLoadAndStore) {
+  SimdFloatX4 fx4;
+  float vals[4] __attribute__((aligned(16))) = {1.0F, 2.0F, 3.0F, 4.0F};
+  float *result = static_cast<float *>(aligned_alloc(16, 4 * sizeof(float)));
+  fx4.load(vals);
+  fx4.store(result);
+  for (size_t i = 0; i < 4; i++) {
+    EXPECT_FLOAT_EQ(result[i], vals[i]);
+  }
+  delete[] result;
+}
+
+TEST(SimdFloatX4, UnalignedLoadAndStore) {
+  SimdFloatX4 fx4;
+  float vals[4]{1.0F, 2.0F, 3.0F, 4.0F};
+  float *result = new float[4];
+  fx4.loadU(vals);
+  fx4.storeU(result);
+  for (size_t i = 0; i < 4; i++) {
+    EXPECT_FLOAT_EQ(result[i], vals[i]);
+  }
+  delete[] result;
+}
+
+TEST(SimdFloatX4, Zero) {
+  SimdFloatX4 fx4;
+  float zeros[4] = {0.0F, 0.0F, 0.0F, 0.0F};
+  float *result = new float[4];
+  fx4.zero();
+  fx4.storeU(result);
+  for (size_t i = 0; i < 4; i++) {
+    EXPECT_FLOAT_EQ(result[i], zeros[i]);
+  }
+  delete[] result;
+}
+
+TEST(SimdFloatX4, Reciprocal) {
+  SimdFloatX4 fx4;
+  float vals[4]{1.0F, 2.0F, 3.0F, 4.0F};
+  float expected[4]{1.0F / 1.0F, 1.0F / 2.0F, 1.0F / 3.0F, 1.0F / 4.0F};
+  float *result = new float[4];
+  fx4.loadU(vals);
+  fx4.reciprocal();
+  fx4.storeU(result);
+  // this isn't super accurate
+  for (size_t i = 0; i < 4; i++) {
+    EXPECT_NEAR(result[i], expected[i], 0.0005);
+  }
+  delete[] result;
+}
+
+TEST(SimdFloatX4, OperatorPlus) {
+  SimdFloatX4 a, b, c;
+  float vals[4]{1.0F, 2.0F, 3.0F, 4.0F};
+  float expected[4]{2.0F, 4.0F, 6.0F, 8.0F};
+  float *result = new float[4];
+  a.loadU(vals);
+  b.loadU(vals);
+  c = a + b;
+  c.storeU(result);
+  for (size_t i = 0; i < 4; i++) {
+    EXPECT_FLOAT_EQ(result[i], expected[i]);
+  }
+  delete[] result;
+}
+
+TEST(SimdFloatX4, OperatorMinus) {
+  SimdFloatX4 a, b, c;
+  float vals[4]{1.0F, 2.0F, 3.0F, 4.0F};
+  float expected[4]{0.0F, 0.0F, 0.0F, 0.0F};
+  float *result = new float[4];
+  a.loadU(vals);
+  b.loadU(vals);
+  c = a - b;
+  c.storeU(result);
+  for (size_t i = 0; i < 4; i++) {
+    EXPECT_FLOAT_EQ(result[i], expected[i]);
+  }
+  delete[] result;
+}
+
+TEST(SimdFloatX4, OperatorMultiply) {
+  SimdFloatX4 a, b, c;
+  float vals[4]{1.0F, 2.0F, 3.0F, 4.0F};
+  float expected[4]{1.0F, 4.0F, 9.0F, 16.0F};
+  float *result = new float[4];
+  a.loadU(vals);
+  b.loadU(vals);
+  c = a * b;
+  c.storeU(result);
+  for (size_t i = 0; i < 4; i++) {
+    EXPECT_FLOAT_EQ(result[i], expected[i]);
+  }
+  delete[] result;
+}
+
+TEST(SimdFloatX4, OperatorDivide) {
+  SimdFloatX4 a, b, c;
+  float vals[4]{1.0F, 2.0F, 3.0F, 4.0F};
+  float expected[4]{0.5F, 1.0F, 3.0F / 2.0F, 2.0F};
+  float *result = new float[4];
+  a.loadU(vals);
+  b.set(2.0F);
+  c = a / b;
+  c.storeU(result);
+  // not very accurate, can we do better?
+  for (size_t i = 0; i < 4; i++) {
+    EXPECT_NEAR(result[i], expected[i], 0.0005);
```

Comment:
Division by 2 is exact (except for [very, very tiny floats that are at the edge of underflow](https://en.wikipedia.org/wiki/Denormal_number)). If this test does not pass with `EXPECT_FLOAT_EQ` then it shouldn't pass.